xls/examples/lfsr.x

```diff
@@ -0,0 +1,518 @@
+////////////////////////////////////////////////////////////////////////////////
+// Linear-feedback shift register (LFSR)
+// A template function to compute the next value of an LFSR counter.
+// Taps can be specified to have varying periods for a given number of bits.
+//
+// For example, lfsr(u5:4, u5:0b10100) computes the value that comes after 4
+// in a 5-bit LFSR with taps on bits 2 and 4.
+////////////////////////////////////////////////////////////////////////////////
+
+fn lfsr<BIT_WIDTH: u32>(current_value: uN[BIT_WIDTH], tap_mask: uN[BIT_WIDTH]) -> uN[BIT_WIDTH] {
+	// Compute the new bit from the taps
+	let new_bit = for (index, xor_bit): (u32, u1) in range(u32:0, BIT_WIDTH) {
+		if tap_mask[index+:u1] == u1:0 {
+			xor_bit
+		} else {
+			xor_bit ^ current_value[index+:u1]
+		}
+	} (u1:0);
+
+	// Kick the high bit and insert the new bit
+	current_value[0:{(BIT_WIDTH - u32:1) as s32}] ++ new_bit
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// A series of maximal LFSRs for different bit widths.
+// Source: https://en.wikipedia.org/wiki/Linear-feedback_shift_register
+////////////////////////////////////////////////////////////////////////////////
+
+fn lfsr2(n: u2) -> u2 {
+	lfsr(n, u2:0b11)
+}
+
+fn lfsr3(n: u3) -> u3 {
+	lfsr(n, u3:0b110)
+}
+
+fn lfsr4(n: u4) -> u4 {
+	lfsr(n, u4:0b1100)
+}
+
+fn lfsr5(n: u5) -> u5 {
+	lfsr(n, u5:0b10100)
+}
+
+fn lfsr6(n: u6) -> u6 {
+	lfsr(n, u6:0b110000)
+}
+
+fn lfsr7(n: u7) -> u7 {
+	lfsr(n, u7:0b1100000)
+}
+
+fn lfsr8(n: u8) -> u8 {
+	lfsr(n, u8:0b10111000)
+}
+
+fn lfsr9(n: u9) -> u9 {
+	lfsr(n, u9:0b100010000)
+}
+
+fn lfsr10(n: u10) -> u10 {
+	lfsr(n, u10:0b1001000000)
+}
+
+fn lfsr11(n: u11) -> u11 {
+	lfsr(n, u11:0b10100000000)
+}
+
+fn lfsr12(n: u12) -> u12 {
+	lfsr(n, u12:0b111000001000)
+}
+
+fn lfsr13(n: u13) -> u13 {
+	lfsr(n, u13:0b1110010000000)
+}
+
+fn lfsr14(n: u14) -> u14 {
+	lfsr(n, u14:0b11100000000010)
+}
+
+fn lfsr15(n: u15) -> u15 {
+	lfsr(n, u15:0b110000000000000)
+}
+
+fn lfsr16(n: u16) -> u16 {
+	lfsr(n, u16:0b1101000000001000)
+}
+
+fn lfsr17(n: u17) -> u17 {
+	lfsr(n, u17:0b10010000000000000)
+}
+
+fn lfsr18(n: u18) -> u18 {
+	lfsr(n, u18:0b100000010000000000)
+}
+
+fn lfsr19(n: u19) -> u19 {
+	lfsr(n, u19:0b1110010000000000000)
+}
+
+fn lfsr20(n: u20) -> u20 {
+	lfsr(n, u20:0b10010000000000000000)
+}
+
+fn lfsr21(n: u21) -> u21 {
+	lfsr(n, u21:0b101000000000000000000)
+}
+
+fn lfsr22(n: u22) -> u22 {
+	lfsr(n, u22:0b1100000000000000000000)
+}
+
+fn lfsr23(n: u23) -> u23 {
+	lfsr(n, u23:0b10000100000000000000000)
+}
+
+fn lfsr24(n: u24) -> u24 {
+	lfsr(n, u24:0b111000010000000000000000)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////////////
+
+#[test]
+fn lfsr7_test() {
```

Comment:
Generated with some scrappy scripts.
Isn't cc_test more cumbersome? You'd need a compiler, while you can develop DSLX algorithms entirely with the interpreter if the test are written in DSLX.

Reply:
it's good to have some simple dslx unittests, but it can also be nice to show validation against a software implementation in `cc_test` (similar to what's done in `//third_party/xls_colors`), the bazel workspace already assume a c++ compiler to build XLS itself ;)

Reply:
OK, so I completely agree it is nice to have cc_test in the context of xls. For example, if the LFSR were in std, it would make sense to add cc_test's and the user could assume the function just works. But, for a sample in the example directory, isn't it better to keep it all in pure DSLX?
Maybe I am biased by my setup, but basically, I develop all my samples in a separate private repo, without any bazel, as if I were a developer to which the binaries (interpreter_main, etc.) had been provided. I don't develop anything inside the xls repo.
I hope it makes sense.

Now, if running those tests is expensive or too long, I could definitely understand we don't want too many of them. The LFSR ones are fast, but for the bitonic sample I have in another PR, the tests do take several seconds.

Reply:
we do have some cc reference implementation of the example, see https://github.com/google/xls/blob/main/xls/examples/crc32_reference.cc

I wish those were actual tests that validate the DSLX implementation, I'm curious what other thinks?
@meheff @cdleary @antmicro/xls

Reply:
The test comparing against a C++ implementation could come in a separate PR.